From e8329e0f34cc5e58f0938fcb93e5d18ca83c0d8f Mon Sep 17 00:00:00 2001
From: "wanwei.jiang" <wanwei.jiang@amlogic.com>
Date: Fri, 4 Mar 2022 15:54:13 +0800
Subject: [PATCH 23/95] freertos: arm32 issue [1/1]

PD#SWPL-74035

Problem:
the definition of macro IPI_FREERTOS is wrong.
(warn_slowpath_fmt) from [<c0f060e8>] (set_smp_ipi_range+0x50/0x110)
(set_smp_ipi_range) from [<c0f2384c>] (__gic_init_bases+0x2cc/0x2f0)

Solution:
IPI_FREERTOS = NR_IPI

Verify:
c2 arm32

Change-Id: I2c1665ee7ea09ed94773ea85863eec41274d2a74
Signed-off-by: wanwei.jiang <wanwei.jiang@amlogic.com>
---
 arch/arm/kernel/smp.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/arch/arm/kernel/smp.c b/arch/arm/kernel/smp.c
index fb3be9b58a1c..9ba5d4d232f3 100644
--- a/arch/arm/kernel/smp.c
+++ b/arch/arm/kernel/smp.c
@@ -74,10 +74,10 @@ enum ipi_msg_type {
 	IPI_CPU_STOP,
 	IPI_IRQ_WORK,
 	IPI_COMPLETION,
+	NR_IPI,
 #if IS_ENABLED(CONFIG_AMLOGIC_FREERTOS)
-	IPI_FREERTOS = 7,
+	IPI_FREERTOS = NR_IPI,
 #endif
-	NR_IPI,
 	/*
 	 * CPU_BACKTRACE is special and not included in NR_IPI
 	 * or tracable with trace_ipi_*
@@ -672,13 +672,11 @@ static void do_handle_IPI(int ipinr)
 		ipi_complete(cpu);
 		break;
 
+	case IPI_CPU_BACKTRACE:
 #if IS_ENABLED(CONFIG_AMLOGIC_FREERTOS)
-	case IPI_FREERTOS:
-		freertos_finish();
-		break;
+		if (!freertos_finish())
+			break;
 #endif
-
-	case IPI_CPU_BACKTRACE:
 		printk_deferred_enter();
 		nmi_cpu_backtrace(get_irq_regs());
 		printk_deferred_exit();
-- 
2.25.1

